airgun/entities/domain.py

```diff
@@ -0,0 +1,192 @@
+
+        self._submit_domain_form(view, **kwargs)
+
+        wait_for(
+            lambda: view.flash.messages, fail_condition=[], timeout=10,
+            delay=2, message='wait for flash success'
+        )
+        view.flash.assert_success_message(self._CREATE_SUCCESSFUL.format(dns_domain))
+
+    def search(self, value=None):
+        """Search for 'value' and return domain names that match.
+
+        :param value: text to filter (default: no filter)
+        """
+        view = self.navigate_to(self, 'All')
+        view.search(value or '')
+        return [row['Description'].text for row in view.table.rows()]
+
+    def read(self, domain_name):
+        """Return dict with properties of domain."""
+        view = self.navigate_to(self, 'Edit', entity_name=domain_name)
+        properties = dict(
+            dns_domain=view.domain.dns_domain.read(),
+            full_name=view.domain.full_name.read(),
+            dns_capsule=view.domain.dns_capsule.read(),
+            parameters=view.parameters.params.read(),
+            locations=view.locations.multiselect.read()['assigned'],
+            organizations=view.organizations.multiselect.read()['assigned']
+        )
+        # Return only non-null entries
+        return {key: val for key, val in properties.items() if val}
+
+    def edit(self, domain_name, **kwargs):
+        """
+        Edit an existing domain.
+
+        Only the kwargs which are passed in will be modified. You must modify
+        at least one value.
+
+        :param domain_name: name of domain to update
+        :param dns_domain: (optional) str, full DNS domain name
+        :param full_name: (optional) str, full name describing the domain
+        :param dns_capsule: (optional) str of DNS capsule name
+        :param parameters: (optional) param dict or list of param dicts to set
+            Format: [{'name': str, 'value': str}]
+            Existing params are overwritten.
+        :param locations: (optional) list of location strings to be assigned.
+            All other locations will be unassigned.
+        :param organizations: (optional) list of org strings to be assigned.
+            All other orgs will be unassigned.
+        """
+        if not kwargs:
+            raise ValueError('no kwargs given for edit()')
+
+        view = self.navigate_to(self, 'Edit', entity_name=domain_name)
+        self._submit_domain_form(view, **kwargs)
+        wait_for(
+            lambda: view.flash.messages, fail_condition=[], timeout=10,
+            delay=2, message='wait for flash success'
+        )
+
+        # If domain is renamed, flash message contains the new name...
+        updated_name = kwargs.get('dns_domain') or domain_name
+        view.flash.assert_success_message(
+            self._UPDATE_SUCCESSFUL.format(updated_name))
+
+    def delete(self, name):
+        view = self.navigate_to(self, 'All')
+        self.search(name)
+        row = view.table.row(('Description', name))
```

Comment:
probably you should define that your table contains button action as third column and access it like that, so it will be on view layer, not on entity one that basically should not care about that at all

Reply:
it's already defined, @bsquizz is just using a bit different style of identifying needed row (which is fine as it's supported & documented way)

Reply:
Correct, if after doing `search` the row with that "Description" is not found, it'll raise a RowNotFound error. So this ensures the row we're looking for after the `search()` is in there.